native caching (#288)
AdobeDocs/graphql-mesh-gateway: src/pages/mesh/advanced/caching/index.md

# Cache-control headers

> API Mesh supports dynamic content caching [natively](#api-mesh-native-caching). You can also [provide your own content delivery network (CDN)](#use-your-own-cdn), such as [Fastly](./fastly.md). Dynamic content caching helps improve site load times and reduces consumption costs associated with bandwidth. A CDN's cache-control headers determine how queried information is cached.

Reviewer: ```suggestion
API Mesh supports dynamic content caching [natively](#api-mesh-native-caching) using standard [HTTP caching](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching). You can also [provide your own content delivery network (CDN)](#use-your-own-cdn), such as [Fastly](./fastly.md). Dynamic content caching helps improve site load times and reduces consumption costs associated with bandwidth. A CDN's cache-control headers determine how queried information is cached.
```

            "endpoint": "https://venia.magento.com/graphql"
          }
        }
      }
    ],
    "responseConfig": {
      "includeHTTPDetails": true,
      "cache": true
    }
  }
}
```

<InlineAlert variant="info" slots="text"/>

Currently, query-level caching is not supported.

<InlineAlert variant="info" slots="text"/>

GET requests are limited to 2,048 characters.

## Response headers

You can control how caching works by modifying the directives within the `cache-control` response headers returned by the sources. At least one source included in the GraphQL query must return a `cache-control` header to consider the entire request for caching. You do not need to explicitly include the `cache-control` header in the source's `responseConfig`.

### Resolving conflicts between sources

When cache-control header values from multiple sources conflict, API Mesh selects the lowest and most restrictive value. The following section explains which values are returned when [cache-control directives](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) conflict.

If no sources in the query return a `cache-control` header, then caching is skipped.

> If you have three sources included in a query, and only one source includes a `cache-control` header, the combined response from all three sources gets cached based on the directives of the source that triggers caching.

Reviewer: ```suggestion
If you have multiple sources included in a query, and only one source includes a `cache-control` header, the combined response from all sources gets cached based on the directives of the source that triggers caching.
```

The `no-store` directive supersedes all other directives. If your source's cache-control headers contain this directive, then the mesh does not return other headers.

<InlineAlert variant="info" slots="text"/>

> You can also use a header value of `x-include-metadata=true` to return all headers.

Reviewer: ```suggestion
You can also include a request header value of `x-include-metadata=true` to return response headers from all sources .
```

> You can also include a request header value of `x-include-metadata=true` to return response headers from all sources .

Reviewer: Sorry, this was probably a typo on my part.

```suggestion
You can also include a request header value of `x-include-metadata=true` to return response headers from all sources.
```

<InlineAlert variant="info" slots="text"/>

> GET requests are limited to 2,048 characters.

Reviewer: This can be moved above after the `Enable caching` example.

Also, the `x-include-metadata` inline alert above relates to response headers. I think we can move it under `How conflicting header values are resolved` as a way to inspect the cache-control headers coming back from each source.

> ###  Response headers

Reviewer: Response headers is a super important aspect of caching. I imagine a logical ordering of sections like:
1. Overview (using cache-control headers)
2. Resolving conflicts between sources
3. Overriding cache-control using mesh-level responseConfig
4. Verifying the caching behavior using response headers

That said, an overview can help re-emphasize the role of response headers in this feature.

e.g. "Caching is controlled by the directives within the `cache-control` response headers returned by the sources. At least one source included in the GraphQL query **_must_** return a cache-control header for the entire request to be considered for caching. The user does not need to explicitly include the `cache-control` header this in the source's `responseConfig`."

The following response headers are returned when caching is enabled:

- `Age` - On cache `HIT`, cached response age in seconds.

- `Cache-Status` - `HIT` or `MISS`.

- `Date` - Date of the response in UTC.

- `Etag` - Unique identifier for a response.

- `Expires` - UTC date when the cached response expires.

> - `Last-Modified` - UTC date when the cached response was stored.

Reviewer: Can we move this to the end, under a section like `Verifying the cached behavior`?

"You can verify the caching behavior of your GraphQL requests based the values of the response headers returned when caching is enabled."

> #### How conflicting header values are resolved

Reviewer 2: One additional case here is that public and private are mutually exclusive. We should describe an example case to support this as well as such:

- Source 1: public, max-age=30, s-maxage=600
- Source 2: private, max-age=60
- Result: private, max-age=30

  - no-store

Example 3

> Public and private headers are mutually exclusive, since `private` is more restrictive, API Mesh selects the values associated with the `private` header.

Reviewer 3: ```suggestion
Public and private headers are mutually exclusive. Since `private` is more restrictive, API Mesh selects the values associated with the `private` header.
```

  - no-store

Example 3

Public and private headers are mutually exclusive, since `private` is more restrictive, API Mesh selects the values associated with the `private` header.

- Source 1 response headers

  - public, max-age=30, s-maxage=600

- Source 2 response headers

  - private, max-age=60

- Combined HTTP response headers

  - private, max-age=30

> ### In the mesh configuration file

Reviewer: Can we rename this section to emphasize the objective? 

Something like `Overriding the cache-control in your mesh`

While we recommend using the [native API Mesh caching](#api-mesh-native-caching) functionality, you can also provide your own content delivery network (CDN), such as Fastly. Refer to the [Fastly caching example](./fastly.md) for more information.

<InlineAlert variant="info" slots="text"/>

When using your own CDN, you must invalidate the cache after modifying a mesh configuration, or you will receive stale information. There is currently no way to manually invalidate your cache, instead you must wait for the Time to Live (`TTL`) to expire on the cached content.

<InlineAlert variant="info" slots="text"/>

`POST` requests are not supported when bringing your own CDN.

## Verifying the caching behavior using response headers

You can verify the caching behavior of GraphQL requests based on the values of the returned response headers when caching is enabled.

**Response headers**

The following response headers are returned when caching is enabled:

- `Age` - On cache `HIT`, cached response age in seconds.

- `Cache-Status` - `HIT` or `MISS`.

> - `Date` - Date of the response in UTC.

Reviewer: Forgot to mention that this header will always be present regardless of caching. While it's a new response header we are introducing, we can remove it from this list.

  }
}
```

> ## Use your own CDN

Reviewer: Can we make this the last section?

Author: good call. done.

  }
}
```

## Verifying the caching behavior using response headers

You can verify the caching behavior of GraphQL requests based on the values of the returned response headers when caching is enabled.

**Response headers**

The following response headers are returned when caching is enabled:

- `Age` - On cache `HIT`, cached response age in seconds.

- `Cache-Status` - `HIT` or `MISS`.

- `Etag` - Unique identifier for a response.

- `Expires` - UTC date when the cached response expires.

- `Last-Modified` - UTC date when the cached response was stored.

## Use your own CDN

To disable native caching in API Mesh and use your own CDN, ensure that your `responseConfig` contains `"cache": false` to avoid double caching.

> While we recommend using the [native API Mesh caching](#api-mesh-native-caching) functionality, you can also provide your own content delivery network (CDN), such as Fastly. Refer to the [Fastly caching example](./fastly.md) for more information.

Reviewer 3: I think this paragraph should be first.

  }
}
```

> ## Use your own CDN

Reviewer 2: Should we call out here that if a user wants to use their own CDN for caching, they should ensure that responseConfig.cache is set to false to avoid double caching?
